### PR TITLE
Image viewer 2d projection

### PR DESCRIPTION
**Pull request by toastisme**
This allows 2d projections to be added to a Detector at the creation stage. 

The purpose of this is to avoid special case checks buried in image_viewer scripts for detectors that do not work with project_2d (i.e FormatCBFMiniPilatusDLS12M). Unique projection behaviour is now defined only in a format class, making it much easier to add more edge case detectors. image_viewer scripts then use explicit projections over project_2d if they are present on the Detector. projection_2d keys are only added to .expt files if they have been manually set, to avoid noise in the majority of cases with normal detectors.

### CODE REVIEW

**Comment by toastisme** [comment]
@rjgildea suggested ensuring the panel projection 2d getter is consistent with what comes out of `project_2d`. I've renamed `project_2d` to `get_detector_projection_2d_axes`, to better describe what it does. I've then abstracted out the equivalent for `panel.get_projection_2d` from `tile_generation.py` into `detector_helpers.py` as `get_panel_projection_2d_from_axes`. This function returns rotation and translation, like `panel.get_projection_2d`.

File: newsfragments/422.misc [hunk 1]
```diff
@@ -0,0 +1,2 @@
+2D projection getters/setters added to Panel.
+Panel 2D projection using axes moved from DIALS to detector_helpers.py and made consistent with panel.get_2d_projection.
```
Comment (dagewa) [inline newsfragments/422.misc:2]: I defer to @ndevenish for comment

File: src/dxtbx/model/detector_helpers.py [hunk 2]
```diff
@@ -11,6 +11,8 @@
 except ImportError:
     sklearn = None
 
+from xfel.cftbx.detector.metrology import get_projection_matrix
```
Comment (dagewa) [inline src/dxtbx/model/detector_helpers.py:14]: We should avoid a dependence on `xfel` if possible as `dxtbx` should remain an upstream dependency for DIALS and cctbx.xfel. That function looks fairly basic so perhaps could just be reproduced here, or the original moved to e.g. `scitbx`

File: src/dxtbx/model/detector_helpers.py [hunk 3]
```diff
@@ -294,9 +296,15 @@ def set_detector_distance(detector, distance):
     detector[0].set_frame(fast_axis, slow_axis, origin)
 
 
-def project_2d(detector):
-    """
-    Project panel origin, fast and slow onto the best-fitting 2D plane.
+def get_detector_projection_2d_axes(detector):
```
Comment (dagewa) [inline src/dxtbx/model/detector_helpers.py:299]: :+1:

**Review by dagewa — CHANGES_REQUESTED** [review]
I tested with an I23 image and that seemed fine. Appreciate the campsite tidying and naming improvements. The main issue is the addition of a dxtbx→xfel dependency

**Comment by toastisme** [comment]
Removed xfel dependency. As it's one line (and we're not even using all of the output of get_projection_matrix) I've just added it explicitly. If get_projection_matrix was moved, there's probably a conversation about having it in numpy instead, where it should be, and making changes to it being used in cctbx, which feels out of scope for this PR.

File: tests/model/test_detector.py [hunk 4]
```diff
@@ -305,11 +309,135 @@ def test_project_2d():
...
+    empty_translation = (0, 0)
+
+    # Valid rotation value
+    rotation = (1, 0, 0, 1)
+
+    ## Test single panel detector
+    detector = create_detector(offset=0)
+
+    # Test detector with no projections set explicity has no 2d projection
+    assert detector.has_projection_2d() is False
```
Comment (rjgildea) [inline tests/model/test_detector.py:408]: See comments above r.e. explicitly comparing to bool.
```suggestion
    assert not detector.has_projection_2d()
```

File: src/dxtbx/model/detector.h [hunk 5]
```diff
@@ -763,6 +763,15 @@ namespace dxtbx { namespace model {
...
     }
+  
+    bool has_projection_2d(){
+      for (std::size_t i = 0; i < size(); ++i){
+        if (!(*this)[i].has_projection_2d()){
+          return false;
+        }
+      }
+      return true;
+    }
```
Comment (rjgildea) [inline src/dxtbx/model/detector.h:774]: Do we need an explicit `Detector.has_projection_2d()` function, particularly if we're making the suggested changes below to remove `Panel.has_projection_2d()` and allow `Panel.get_projection_2d()` to optionally return `None`. Then you would just be able to do e.g. `all([p.get_projection_2d() for p in detector])` in Python, but with more control over what to do if only _some_ of the panels have a 2d projection.

File: src/dxtbx/model/detector_helpers.py [hunk 6]
```diff
@@ -294,9 +294,15 @@ def set_detector_distance(detector, distance):
...
+
+    """Project panel origins, fast and slow axes onto the best-fitting 2D plane.
+
+    :param detector: detector used to generate axes
+    :type detector: Detector
+
+    :return: 2D origin, fast, and slow vectors for each panel
+    :rtype: list(tuple(float, float)), list(tuple(float, float)), list(tuple(float, float))
```
Comment (rjgildea) [inline src/dxtbx/model/detector_helpers.py:305]: As below, I think type annotations would be more modern/useful here than this style of docstrings.

**Review by rjgildea — CHANGES_REQUESTED** [review]
Thank you for adding this, it will be a good addition to remove hardcoded cases for the P12M in the dials.image_viewer, as well as enabling support for similarly non-flat detectors.

I think the C++ interface could potentially be cleaned up and made more "Pythonic" by allowing the projections to be `None` using `boost::optional`, and it would probably be more useful to use type annotations rather than docstrings to indicate parameter types (although I appreciate this is still better than nothing at all).

**Comment by toastisme** [comment]
@rjgildea thanks for the detailed feedback. Had not come across boost::optional before. I think I've applied all the suggestions. The only thing I've left is the `Detector.has_projection_2d()`. It's largely syntactic sugar but I also included it to avoid partially set projections. Is there a use case for setting some panels manually, and others using the beam centre (@dagewa)? That feels like it would be a mess to me, but happy to change it if it's viable.

File: src/dxtbx/model/detector_helpers.py [hunk 7]
```diff
@@ -294,15 +300,10 @@ def set_detector_distance(detector, distance):
     detector[0].set_frame(fast_axis, slow_axis, origin)
 
 
-def get_detector_projection_2d_axes(detector):
-
-    """Project panel origins, fast and slow axes onto the best-fitting 2D plane.
-
-    :param detector: detector used to generate axes
-    :type detector: Detector
+def get_detector_projection_2d_axes(detector: Detector) -> Tuple[Tuple, Tuple, Tuple]:
```
Comment (rjgildea) [inline src/dxtbx/model/detector_helpers.py:303]: You could define a [type alias](https://docs.python.org/3/library/typing.html#type-aliases) to make the return signature slightly clearer:
```suggestion
Coord2d = Tuple[float, float]

def get_detector_projection_2d_axes(detector: Detector) -> Tuple[Coord2d, Coord2d, Coord2d]:
```

File: src/dxtbx/model/detector_helpers.py [hunk 8]
```diff
@@ -410,25 +411,16 @@ def get_detector_projection_2d_axes(detector):
...
 
 def get_panel_projection_2d_from_axes(
-    panel, image_data, fast_axis_2d, slow_axis_2d, origin_2d
-):
+    panel: Panel,
+    image_data: matrix,
+    fast_axis_2d: matrix.col,
+    slow_axis_2d: matrix.col,
+    origin_2d: matrix.col,
+) -> Tuple[Tuple, Tuple]:
```
Comment (rjgildea) [inline src/dxtbx/model/detector_helpers.py]: As above, perhaps better naming could be something like:
```
Float2 = Tuple[float, float]
Float4 = Tuple[float, float, float, float]

def get_panel_projection_2d_from_axes(...) -> Tuple[Float4, Float2]:
    ...
```

**Comment by dagewa** [comment]
> @rjgildea thanks for the detailed feedback. Had not come across `boost::optional` before. I think I've applied all the suggestions. The only thing I've left is the `Detector.has_projection_2d()`. It's largely syntactic sugar but I also included it to avoid partially set projections. Is there a use case for setting some panels manually, and others using the beam centre (@dagewa)? That feels like it would be a mess to me, but happy to change it if it's viable.

I can't think of a reason to only set some panels. When we're doing this, we want to do it to the whole detector

**Pull request merged**